### PR TITLE
feat: lint sources

### LINKED ISSUE #76: feature request: enable linting of sources
This is a super cool tool!

Is it on the roadmap to implement dbt sources as a lint-able node type?

I'd be happy to contribute a PR if that's something worthwhile to this project.

### PR DESCRIPTION
**Pull request by otosky**
# Overview
Extends rules so that they can be used to against dbt [sources](https://docs.getdbt.com/docs/build/sources) in addition to models.

# Usage
A rule defines what resource-type it acts against in the type signature of the function it wraps or in a class-based `evaluate` method:

```python
from dbt_score import Model, Source rule, Rule, RuleViolation

# decorator-based
# for a Model
@rule
def model_has_description(model: Model) -> RuleViolation | None:
    """A model should have a description."""
    if not model.description:
        return RuleViolation(message="Model lacks a description.")

# for a Source
@rule
def has_description(source: Source) -> RuleViolation | None:
    """A source should have a loader defined."""
    if not source.loader:
        return RuleViolation(message="Source lacks a loader.")

# class-based
class ExampleSource(Rule):
    """Example class-based rule."""

    description = "A source should have a loader defined."

    def evaluate(self, source: Source) -> RuleViolation | None:
        """Evaluate source."""
        if not source.loader:
            return RuleViolation(message="Source lacks a loader.")
```

The `Evaluation` handler is then responsible for applying source-rules to Source objects and model-rules to Model objects.

---

closes #76

### CODE REVIEW

**Comment by otosky** [comment]
woops thought that I opened this PR against my own fork - will move out of draft when I've added the rest of the feature :slightly_smiling_face:

**Comment by matthieucan** [comment]
> woops thought that I opened this PR against my own fork - will move out of draft when I've added the rest of the feature 🙂

No problem, sounds good!

File: src/dbt_score/rule.py
Comment (otosky) [inline src/dbt_score/rule.py:99]: Since the type-annotation is critical to a proper resource-type discriminator being set on the class, I thought that it was valid to throw an error here if there isn't a param typed as `Source` or `Model`.

**Comment by otosky** [comment]
@matthieucan - let me know if you think this is heading in the right direction

File: src/dbt_score/evaluation.py
Comment (jochemvandooren) [inline src/dbt_score/evaluation.py:45]:
```suggestion
        # For each evaluable, its results
```

File: src/dbt_score/evaluation.py
Comment (jochemvandooren) [inline src/dbt_score/evaluation.py:48]:
```suggestion
        # For each evaluable, its computed score
```

File: src/dbt_score/evaluation.py
Comment (jochemvandooren) [inline src/dbt_score/evaluation.py:58]:
```suggestion
        for evaluable in chain(self._manifest_loader.models, self._manifest_loader.sources):
```

Comment (jochemvandooren) [reply src/dbt_score/evaluation.py:58]: Also for further occurrences in this function, let's try to get rid of all the `model` references where now `sources` are included

File: src/dbt_score/__init__.py
Comment (jochemvandooren) [inline src/dbt_score/__init__.py:3]: We have to adapt `dbt_score.model_filter` in a similar way as `dbt_score.rule` was changed in order to apply filters to sources as well

**Review by jochemvandooren — COMMENTED** [review]
I think this looks very good already! :rocket: Just a couple of things we need to think about:

- Filters cannot be applied to sources now
- Code has `model` references almost everywhere (oops :sweat_smile:, that was me probably), we should change that. E.g. `scorer.py` even has methods like `score_model` that will now be applied to `sources` as well.
- Can we distinguish source and model in the formatters?
- Needs some documentation in `/docs`!

Let me know what you think about it, I am happy to help if needed!

File: src/dbt_score/__init__.py
Comment (otosky) [reply src/dbt_score/__init__.py:3]: Updated in fe80bdf!  I will update the argument name references here as well in another commit.

How would you like to address the naming of the class/decorator now that it applies to both `Model`s and `Source`s?  Is it okay to have a breaking change here to rename it `RuleFilter` or `EvaluableFilter` or something else?

**Comment by otosky** [comment]
> * Can we distinguish source and model in the formatters?

@jochemvandooren 

For the human-readable formatter, my thought was to prefix models with `M:` and sources with `S:`, but let me know if you have a different idea.

Example:
```
🥇 M:model1 (score: 10.0)
    OK   tests.conftest.rule_severity_low
    ERR  tests.conftest.rule_severity_medium: Oh noes
    WARN (critical) tests.conftest.rule_severity_critical: Error

🥇 S:source1.table1 (score: 10.0) 
    OK   tests.conftest.rule_severity_low
    ERR  tests.conftest.rule_severity_medium: Oh noes
    WARN (critical) tests.conftest.rule_severity_critical: Error
```

File: src/dbt_score/__init__.py
Comment (jochemvandooren) [reply src/dbt_score/__init__.py:3]: Nice! :raised_hands: I will have a look at the changes. 

I am afraid that this indeed warrants a breaking change, the name does not make sense anymore. But, breaking changes are expected considering we are still at version `0.6`. I like `RuleFilter` the most!

**Comment by jochemvandooren** [comment]
> > * Can we distinguish source and model in the formatters?
> 
> @jochemvandooren
> 
> For the human-readable formatter, my thought was to prefix models with `M:` and sources with `S:`, but let me know if you have a different idea.
> 
> Example:
> 
> ```
> 🥇 M:model1 (score: 10.0)
>     OK   tests.conftest.rule_severity_low
>     ERR  tests.conftest.rule_severity_medium: Oh noes
>     WARN (critical) tests.conftest.rule_severity_critical: Error
> 
> 🥇 S:source1.table1 (score: 10.0) 
>     OK   tests.conftest.rule_severity_low
>     ERR  tests.conftest.rule_severity_medium: Oh noes
>     WARN (critical) tests.conftest.rule_severity_critical: Error
> ```

I like the idea, keeps it concise! :+1:

File: src/dbt_score/models.py
Comment (jochemvandooren) [reply src/dbt_score/models.py:413]: Same for `_select_models`

**Review by jochemvandooren — COMMENTED** [review]
This has become a huge PR with all the renaming and stuff, great effort :raised_hands: I have played around with the code locally and everything works perfectly! Just left some comments about getting rid of all mentions of `model` in the code.

Also, all of the documentation needs to be updated as well, I am happy to assist with this one, please let me know!

File: src/dbt_score/cli.py
Comment (otosky) [inline src/dbt_score/cli.py:91]: @jochemvandooren 

Were you okay with this change in configuration?  I think this would be my preference (a single setting to fail any item under a threshold), but I could see the argument for separate settings for models or sources as well.

**Comment by otosky** [comment]
Thanks for a thorough review @jochemvandooren !  Will address remaining feedback, update the formatters, and take a first stab at some of the docs shortly!  I will definitely take up your offer for support on the docs.

File: src/dbt_score/cli.py
Comment (matthieucan) [reply src/dbt_score/cli.py:91]: While there, could we use dashes `-` instead of underscores `_` in the CLI option? This was a mistake when introduced 🙏

Comment (matthieucan) [reply src/dbt_score/cli.py:91]: "evaluable" seems very correct and fancy, but perhaps not the best for the CLI. Maybe "entity" could work better, as in dbt entities are models, sources, etc. WDYT?

File: src/dbt_score/models.py
Comment (matthieucan) [inline src/dbt_score/models.py:1]: Should this file be renamed accordingly to something like `dbt_entities.py`?

**Review by matthieucan — COMMENTED** [review]
Incredible work @otosky !
I played a bit with it and it works fine! Very much looking forward to have this 💪

File: src/dbt_score/models.py
Comment (otosky) [reply src/dbt_score/models.py:401]: Updated!

File: src/dbt_score/models.py
Comment (otosky) [reply src/dbt_score/models.py:423]: Updated `_select_models` -> `_filter_evaluables`.

Note that the selector name output by `dbt ls --output name` for sources is of the form `source_name.source_table_alias`.

FWIW, I have some thoughts for another leg of work to add some better integration testing here - all the mocking/stubbing in the manifest related codepaths didn't leave me feeling confident that the changes I was making would actually work on a real project.

Comment (otosky) [reply src/dbt_score/models.py:425]: Updated `_reindex_tests` - it needed a little massaging to fit sources.

File: src/dbt_score/rule_filter.py
Comment (otosky) [reply src/dbt_score/rule_filter.py:64]: updated

File: src/dbt_score/evaluation.py
Comment (otosky) [reply src/dbt_score/evaluation.py:63]: This is actually a different check to match up the resource_type of the rule to the resource_type of the model.  But based on that confusion, I think it's probably better to push that down into the `rule.should_evaluate` method?

File: src/dbt_score/cli.py
Comment (otosky) [reply src/dbt_score/cli.py]: updated

File: src/dbt_score/evaluation.py
Comment (otosky) [reply src/dbt_score/evaluation.py]: updated in 9b5af62

File: src/dbt_score/models.py
Comment (otosky) [reply src/dbt_score/models.py:1]: IMO `models` is idiomatic enough for it's contents that I don't think users will get too tripped up with the mixed terminology.  I don't feel strongly either way, though!

File: docs/create_rules.md
Comment (otosky) [inline docs/create_rules.md:141]: I'm realizing now that I haven't actually made any changes to enforce that resource_types match across `rule_filter`s attached to `rule`s.  

This would be easy to account for by addressing it in the `rule.should_evaluate` method, but would it be better to throw an error earlier with a nice message such that something like this can't be configured at all?

```python
    @rule_filter
    def source_schema_in_x(source: Source):
        """Only keep sources in schema x."""
        return schema_in_x(source)

    @rule(rule_filters={source_schema_in_x()})
    def model_has_description(model: Model):
        """Validates that a model has a description."""
        if not model.description:
            return RuleViolation("Model does not have a description.")
```

File: src/dbt_score/models.py
Comment (jochemvandooren) [reply src/dbt_score/models.py:1]: I came to the same conclusion @otosky , no strong opinion either!

File: docs/create_rules.md
Comment (jochemvandooren) [reply docs/create_rules.md:141]: I agree throwing an error earlier makes more sense indeed, it improves UX in my opinion! :+1:

**Comment by jochemvandooren** [comment]
> Thanks for a thorough review @jochemvandooren ! Will address remaining feedback, update the formatters, and take a first stab at some of the docs shortly! I will definitely take up your offer for support on the docs.

Next week I will help you on the docs and further review the PR, thanks for all the amazing work already :raised_hands:

File: src/dbt_score/models.py
Comment (matthieucan) [reply src/dbt_score/models.py:1]: Alright! 👍

File: docs/create_rules.md
Comment (otosky) [reply docs/create_rules.md:141]: addressed in 0eeae68

File: src/dbt_score/evaluation.py
Comment (jochemvandooren) [reply src/dbt_score/evaluation.py]: Nice, thanks!

File: docs/create_rules.md
Comment (jochemvandooren) [reply docs/create_rules.md:141]: Great :raised_hands:

**Comment by jochemvandooren** [comment]
> > Thanks for a thorough review @jochemvandooren ! Will address remaining feedback, update the formatters, and take a first stab at some of the docs shortly! I will definitely take up your offer for support on the docs.
> 
> Next week I will help you on the docs and further review the PR, thanks for all the amazing work already 🙌

@otosky The code looks perfect! As promised I did a final check and tried to find all occurrences of `model` and replaced it by something more appropriate :mag: https://github.com/PicnicSupermarket/dbt-score/pull/78/commits/d595ea278881b062ca5613b6fd3d190ecc048fab

I also added a `CHANGELOG.md` entry already, feel free to improve of course. Final thing that's remaining is some linting errors on lines exceeding line length. Once those are fixed I am ready to approve :rocket:

**Comment by otosky** [comment]
Thank you @jochemvandooren!  

Fixed the line lengths in a2c0924 and made some tweaks to the changelog in f88aa7c.

One final Q:  I made use of two functions from [more-itertools](https://more-itertools.readthedocs.io/en/stable/) - [first](https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.first) and [first_true](https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.first_true).  `more-itertools` came as a sub-dependency of `dbt`, which I see has now been made a dev-dep.  Would you like me to rewrite/vendor the 2 functions I used above?  Or add `more-itertools` as proper dependency?

**Comment by jochemvandooren** [comment]
> Thank you @jochemvandooren!
> 
> Fixed the line lengths in [a2c0924](https://github.com/PicnicSupermarket/dbt-score/commit/a2c09242ec2e8e80ee85c37e6e87d80c7cd758bd) and made some tweaks to the changelog in [f88aa7c](https://github.com/PicnicSupermarket/dbt-score/commit/f88aa7c249ee7b0e6b031c80c331bd6617658a94).
> 
> One final Q: I made use of two functions from [more-itertools](https://more-itertools.readthedocs.io/en/stable/) - [first](https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.first) and [first_true](https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.first_true). `more-itertools` came as a sub-dependency of `dbt`, which I see has now been made a dev-dep. Would you like me to rewrite/vendor the 2 functions I used above? Or add `more-itertools` as proper dependency?

Ah, good point! If it can be prevented, I'd like to keep the number of dependencies low. But I can imagine having to rewrite the function is a hassle, I will leave it up to you!

**Comment by otosky** [comment]
I went the vendoring route, since it's really just the function `first_true` that is nicer syntax sugar and used in more than one place.  :+1:

**Comment by matthieucan** [comment]
> I went the vendoring route, since it's really just the function `first_true` that is nicer syntax sugar and used in more than one place. 👍

I believe `more_itertools.py` was not committed?

**Comment by otosky** [comment]
> > I went the vendoring route, since it's really just the function `first_true` that is nicer syntax sugar and used in more than one place. 👍
> 
> I believe `more_itertools.py` was not committed?

:facepalm: totally right - just pushed it up!

**Review by jochemvandooren — APPROVED** [review]
Amazing :star_struck: , just needs a rebase on master!

File: CHANGELOG.md
Comment (matthieucan) [inline CHANGELOG.md:11]: That's a feature, not a breaking change, right?

File: src/dbt_score/rule_filter.py
Comment (matthieucan) [inline src/dbt_score/rule_filter.py:1]:
```suggestion
"""Evaluable filtering to choose when to apply specific rules."""

```

Comment (matthieucan) [reply src/dbt_score/rule_filter.py:1]: (adding empty line)

**Review by matthieucan — APPROVED** [review]
Minor nitpicks.

Incredible work @otosky, I'm really keen to start using this feature. Thanks a lot! 💪

File: CHANGELOG.md
Comment (otosky) [reply CHANGELOG.md:11]: Yeah, I'll fix that

File: src/dbt_score/formatters/manifest_formatter.py
Comment (otosky) [inline src/dbt_score/formatters/manifest_formatter.py:31]: Noticing after taking another pass at the diff that sources need to be pushed into `manifest["sources"]` instead of `manifest["nodes"]`.  Let me fix that quickly.

Comment (matthieucan) [reply src/dbt_score/formatters/manifest_formatter.py:31]: Oh good one, we should have a test for that

Comment (otosky) [reply src/dbt_score/formatters/manifest_formatter.py]: updated in cc6b707

File: src/dbt_score/formatters/manifest_formatter.py
Comment (otosky) [inline src/dbt_score/formatters/manifest_formatter.py:38]: I did the simple thing here and didn't account for any future situation where an Evaluable is something outside of Model | Source, but let me know if it's better to be more defensive here.  Open to suggestions.

Comment (jochemvandooren) [reply src/dbt_score/formatters/manifest_formatter.py:38]: I think fine for now! :ok_hand:

**Comment by jochemvandooren** [comment]
It seems `mypy` was upgraded when rebasing, and it introduced some new linting errors :sweat_smile: I can help you fix them if you would like! Let me know

**Comment by otosky** [comment]
@jochemvandooren I'll take an initial stab at it!

**Comment by otosky** [comment]
There are still a bunch of Liskov violations being raised from mypy that I'm not entirely sure how to fix without having to introduce Generics into the API.

```
tests/test_rule.py:53: error: Argument 1 of "evaluate" is incompatible with supertype "Rule"; supertype defines the argument type as "Model | Source"  [override]
tests/test_rule.py:53: note: This violates the Liskov substitution principle
tests/test_rule.py:53: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
```

Is there any other workaround besides adding an ignore here?  This essentially means that mypy will fail for any downstream users if they use the class-based workflow while developing their rules.

**Comment by jochemvandooren** [comment]
> There are still a bunch of Liskov violations being raised from mypy that I'm not entirely sure how to fix without having to introduce Generics into the API.
> 
> ```
> tests/test_rule.py:53: error: Argument 1 of "evaluate" is incompatible with supertype "Rule"; supertype defines the argument type as "Model | Source"  [override]
> tests/test_rule.py:53: note: This violates the Liskov substitution principle
> tests/test_rule.py:53: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
> ```
> 
> Is there any other workaround besides adding an ignore here? This essentially means that mypy will fail for any downstream users if they use the class-based workflow while developing their rules.

I also spent some time looking into this, and there's no easy solution indeed :disappointed: Considering this will only affect the class-based rules, I suggest we add an ignore for now. I am aware it will introduce the warnings downstream, which isn't ideal. To solve this in a nice way, we might need to restructure some things if we want to keep the API as is, so we might consider this in a follow-up PR. What do you think? I think it's the most pragmatic option!

**Comment by otosky** [comment]
> I also spent some time looking into this, and there's no easy solution indeed 😞 Considering this will only affect the class-based rules, I suggest we add an ignore for now. I am aware it will introduce the warnings downstream, which isn't ideal. To solve this in a nice way, we might need to restructure some things if we want to keep the API as is, so we might consider this in a follow-up PR. What do you think? I think it's the most pragmatic option!

makes sense to me @jochemvandooren! - updated in 0732e15

**Comment by otosky** [comment]
@jochemvandooren @matthieucan thanks again for all your assistance!

**Pull request merged**
